### ISSUE TITLE
Adaptation of engine control commands to the provider interface

### ISSUE THREAD

**Issue by Voev** [openssl/openssl#12503]
Hello!

Is there a correct way to adapt the control commands of the OpenSSL ENGINE API in version 1.1.1 for new version of OpenSSL 3.0 (alpha 5)?

For example, earlier in version 1.1.1, the engine was initialized using a configuration file, and it contains custom commands that were registered by the engine itself (Сommands set the necessary parameters, for example, in the case of PKСS#11, these are the parameters for setting the path of the PKCS#11 module, the token slot number, etc.) Therefore, its setting in the configuration file looks like this:
```
openssl_conf = openssl_def

[openssl_def]
engines = engine_section

[engine_section]
engine_name = engine_conf_section

[engine_conf_section]
CUSTOM_CMD1  = CUSTOM_PARAM1
CUSTOM_CMD2  = CUSTOM_PARAM2
default_algorithms  = ALL
```
In the new version, the provider (rewritten engine) stores the command values ​​in its structure in the `parameters` field (`ossl_provider_st`), which can be obtained in the provider initialization function `OSSL_provider_init` by calling `OSSL_FUNC_core_get_params`. In terms of the syntax of the config file, nothing actually changes:

```
openssl_conf  = openssl_def

[openssl_def]
providers = provider_section

[provider_section]
provider_name = provider_conf_section

[provider_conf_section]
CUSTOM_CMD1  = CUSTOM_PARAM1
CUSTOM_CMD2  = CUSTOM_PARAM2
activate = 1
```

So if a configuration file is used this leads to the fact that the processing of control commands occurs in `OSSL_provider_init`, after which the provider can be used in cryptographic operations. How correct is the described way of processing control commands?

Also passing the required parameters without using the config file is not so obvious. Is there any manual or documentation on how to get a code equivalent to the provider API for a specific engine control command (for example, ```ENGINE_ctrl_cmd_string(e, "CUSTOM_CMD", "CUSTOM_PARAM", 0)```)?

**Comment by levitte** [comment]
You're right, there's no easy way to do the same thing programmatically with providers. That should be terribly difficult to do, though.

Now, engine had pre-ctrls and post-ctrls. I assume that the latter is the more interesting...

**Comment by Voev** [comment]
Thanks for answer

**Comment by levitte** [comment]
If you don't mind, I'll keep this open a bit of time. I have some fixes in mind...

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
